rotkehlchen/chain/ethereum/modules/weth/decoder.py

```diff
@@ -0,0 +1,180 @@
+WETH_WITHDRAW_TOPIC = b'\x7f\xcfS,\x15\xf0\xa6\xdb\x0b\xd6\xd0\xe08\xbe\xa7\x1d0\xd8\x08\xc7\xd9\x8c\xb3\xbfrh\xa9[\xf5\x08\x1be'  # noqa: E501
+
+
+class WethDecoder(DecoderInterface):
+    def __init__(
+            self,
+            ethereum_manager: 'EthereumManager',
+            base_tools: 'BaseDecoderTools',
+            msg_aggregator: 'MessagesAggregator',
+    ) -> None:
+        super().__init__(ethereum_manager, base_tools, msg_aggregator)
+        self.base_tools = base_tools
+        self.weth = A_WETH.resolve_to_evm_token()
+        self.eth = A_ETH.resolve_to_crypto_asset()
+
+    def _decode_weth(
+            self,
+            tx_log: EthereumTxReceiptLog,
+            transaction: EvmTransaction,
+            decoded_events: List[HistoryBaseEntry],
+            all_logs: List[EthereumTxReceiptLog],
+            action_items: Optional[List[ActionItem]],
+    ) -> Tuple[Optional[HistoryBaseEntry], Optional[ActionItem]]:
+        if tx_log.topics[0] == WETH_DEPOSIT_TOPIC:
+            return self._decode_deposit_event(
+                tx_log=tx_log,
+                transaction=transaction,
+                decoded_events=decoded_events,
+                all_logs=all_logs,
+                action_items=action_items,
+            )
+
+        if tx_log.topics[0] == WETH_WITHDRAW_TOPIC:
+            return self._decode_withdrawal_event(
+                tx_log=tx_log,
+                transaction=transaction,
+                decoded_events=decoded_events,
+                all_logs=all_logs,
+                action_items=action_items,
+            )
+
+        return None, None
+
+    def _decode_deposit_event(
+            self,
+            tx_log: EthereumTxReceiptLog,
+            transaction: EvmTransaction,
+            decoded_events: List[HistoryBaseEntry],
+            all_logs: List[EthereumTxReceiptLog],  # pylint: disable=unused-argument
+            action_items: Optional[List[ActionItem]],  # pylint: disable=unused-argument
+    ) -> Tuple[Optional[HistoryBaseEntry], Optional[ActionItem]]:
+        depositor = hex_or_bytes_to_address(tx_log.topics[1])
+        deposited_amount_raw = hex_or_bytes_to_int(tx_log.data[:32])
+        deposited_amount = asset_normalized_value(amount=deposited_amount_raw, asset=self.eth)
+
+        in_event = None
+        # when weth is interacting with a protocol e.g. Uniswap, the protocol handles the
+        # wrapping and unwrapping of WETH
+        is_interacting_with_protocol = False
+        for event in decoded_events:
+            if (
+                event.event_type == HistoryEventType.SPEND and
+                event.counterparty in (WETH_CONTRACT, depositor) and
+                event.balance.amount == deposited_amount and
+                event.asset == self.eth
+            ):
+                if event.counterparty == depositor:
+                    is_interacting_with_protocol = True
+                else:
+                    in_event = event
```

Comment:
this name confused me. This is the **out event**. You are sending a token outside (the ETH you deposit)